Sources/Support/StoreMessagesHelper.swift

```diff
@@ -0,0 +1,195 @@
+        }
+        // Compilation fails in older xcode versions with a similar issue to https://github.com/apple/swift/issues/57379
+        #if swift(>=5.8)
+        Task(priority: .medium) { [weak self] in
+            guard let storeMessagesProvider = self?.storeMessagesProvider else {
+                return
+            }
+            for try await message in storeMessagesProvider.messages {
+                await MainActor.run { [weak self] in
+                    self?.deferredMessages.append(message)
+                }
+            }
+        }
+        #else
+        Logger.error("Defer messages are not enabled when compiling using Xcode <14.2")
+        #endif
+    }
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    @MainActor
+    func showStoreMessages(types: Set<StoreMessageType>) {
+        for message in self.deferredMessages {
+            if let messageType = message.reason.messageType,
+               types.contains(messageType) {
+                    do {
+                        try message.display(in: self.systemInfo.currentWindowScene)
+                    } catch {
+                        Logger.error("Error displaying StoreKit message: \(error)")
+                    }
+            }
+        }
+    }
+
+    #endif
+}
+
+// @unchecked because:
+// - Class is not `final` (it's mocked). This implicitly makes subclasses `Sendable` even if they're not thread-safe.
+// - It has mutable `_deferredMessages` which is necessary due to the availability annotations.
+extension StoreMessagesHelper: @unchecked Sendable {}
+
+protocol StoreMessagesProvider {
+
+    #if os(iOS) || targetEnvironment(macCatalyst) || VISION_OS
+
+    @available(iOS 16.0, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    var messages: any StoreMessageAsyncSequence { get }
+
+    #endif
+}
+
+protocol StoreMessage: Sendable {
+
+    #if os(iOS) || targetEnvironment(macCatalyst) || VISION_OS
+
+    @available(iOS 16.0, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    var reason: Message.Reason { get }
+
+    @available(iOS 16.0, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    @MainActor func display(in scene: UIWindowScene) throws
+
+    #endif
+}
+
+#if os(iOS) || targetEnvironment(macCatalyst) || VISION_OS
+
+@available(iOS 13.0, *)
+@rethrows protocol StoreMessageAsyncIteratorProtocol: AsyncIteratorProtocol where Element == StoreMessage { }
+
+@available(iOS 13.0, *)
+@rethrows protocol StoreMessageAsyncSequence: AsyncSequence where AsyncIterator: StoreMessageAsyncIteratorProtocol { }
+
+@available(iOS 16.0, *)
+@available(macOS, unavailable)
+@available(watchOS, unavailable)
+@available(tvOS, unavailable)
+private final class StoreMessageWrapper: StoreMessage {
```

Comment:
I don't think this is needed, can you make `Message` conform to `StoreMessage` instead?